### PR TITLE
Service cpus limit

### PR DESCRIPTION
**Pull request by jakolehm**
> Specify how much of the available CPU resources a service instance can use. For example, if the host machine has two CPUs and you set `cpus: 1.5`, the service instance will be guaranteed to be able to access at most one and a half of the CPUs. Kontena scheduler will automatically target deployment to nodes that have at least number of specified CPUs available.

Related #2450

### CODE REVIEW

**Comment by jakolehm** [comment]
@SpComb PTAL.

File: server/app/services/scheduler/filter/cpu.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+      # @return [Array<HostNode>]
+      def for_service(service, instance_number, nodes)
+        return nodes if service.cpus.nil?
+
+        candidates = nodes.dup.select { |n|
+          n.cpus.to_f >= service.cpus.to_f
+        }
+
+        if candidates.empty?
+          raise Scheduler::Error, "Did not find any nodes with #{service.cpus} CPUs available"
```
Comment (SpComb) [inline server/app/services/scheduler/filter/cpu.rb]: The filtering logic is also a little dubious... `--cpus 4` doesn't imply that the service needs at least four CPUs to run... there's no reservation of CPU resources that would imply that the CPUs are somehow "available" for the service.

If the host node has other CPU-hungry services scheduled, then it won't get it's four CPUs anyways.

**Comment by SpComb** [comment]
@jakolehm waiting for `status/merge`, unless you still want to address the minor comments.

File: cli/lib/kontena/cli/services/create_command.rb [hunk 2]
```diff
@@ -69,6 +70,7 @@ def parse_service_data_from_options
       data[:volumes_from] = volumes_from_list unless volumes_from_list.empty?
       data[:memory] = parse_memory(memory) if memory
       data[:memory_swap] = parse_memory(memory_swap) if memory_swap
+      data[:cpus] = cpus.to_f if cpus
```
[the remark this answers is not in the document]
Comment (jakolehm) [reply cli/lib/kontena/cli/services/create_command.rb]: Good point, fixed.

File: server/app/services/scheduler/filter/cpu.rb (hunk 1 above)
Comment (jakolehm) [reply server/app/services/scheduler/filter/cpu.rb]: >  guaranteed to be able to access

Comment (SpComb) [reply server/app/services/scheduler/filter/cpu.rb]: What do you mean? If you're talking about the [`docker run`](https://docs.docker.com/engine/admin/resource_constraints/#configure-the-default-cfs-scheduler) docs, then your quote is very much out of context:

> the container will be guaranteed to be able to access ***at most*** one and a half of the CPUs

File: cli/lib/kontena/cli/services/create_command.rb [hunk 3]
```diff
@@ -16,6 +16,9 @@ class CreateCommand < Kontena::Command
     option ["-v", "--volume"], "VOLUME", "Add a volume or bind mount it from the host", multivalued: true
     option "--volumes-from", "VOLUMES_FROM", "Mount volumes from another container", multivalued: true
     option ["-a", "--affinity"], "AFFINITY", "Set service affinity", multivalued: true
+    option "--cpus", "CPUS", "Number of CPUs" do |cpus|
+      Float(cpus)
```
Comment (SpComb) [inline cli/lib/kontena/cli/services/create_command.rb]: :+1 `ERROR: option '--cpus': invalid value for Float(): "five"`

**Pull request merged**